### PR TITLE
[WIP] Run eagle with full cudagraph

### PR DESCRIPTION
**Pull request by zixi-qi**
WIP change to support running v1 eagle speculative decoding with full cudagraph. Currently there is a numerical gap when full cudagraph is turned on:

- piecewise
```
VLLM_USE_V1=1 python examples/offline_inference/eagle.py  --num_spec_tokens 7 --max_num_seqs 1 --num_prompts 1

--------------------------------------------------
mean acceptance length: 2.82
--------------------------------------------------
acceptance at token 0:0.77
acceptance at token 1:0.51
acceptance at token 2:0.28
acceptance at token 3:0.13
acceptance at token 4:0.05
acceptance at token 5:0.03
acceptance at token 6:0.03
```
- full cudagraph
```
VLLM_USE_V1=1 python examples/offline_inference/eagle.py  --num_spec_tokens 7 --max_num_seqs 1 --num_prompts 1 --compilation_config '{"full_cuda_graph": true}'

--------------------------------------------------
mean acceptance length: 2.08
--------------------------------------------------
acceptance at token 0:0.63
acceptance at token 1:0.28
acceptance at token 2:0.11
acceptance at token 3:0.05
acceptance at token 4:0.02
acceptance at token 5:0.00
acceptance at token 6:0.00
```

### CODE REVIEW

**Review by houseroad — COMMENTED** [review]
yeah, there could be some case the CUDA graph handle things incorrectly. @yinghai hit similar problems before.

**Comment by zixi-qi** [comment]
> yeah, there could be some case the CUDA graph handle things incorrectly. @yinghai hit similar problems before.

I think most probably there is something wrong with how draft model's attention metadata is being captured and updated during replay here. I have tried to do some manipulations but no luck so far

File: vllm/v1/spec_decode/eagle.py [hunk 1]
```diff
@@ -169,7 +171,7 @@ def propose(
         self.positions[:num_tokens] = target_positions
         self.hidden_states[:num_tokens] = target_hidden_states
 
-        with set_forward_context(per_layer_attn_metadata,
+        with set_forward_context(None,
```
Comment (zixi-qi) [inline vllm/v1/spec_decode/eagle.py]: AFAIK, in full cudagraph mode, the attention metdata passed into context manager should not have an impact on the results since no python code is executed. However changing the code here does have an impact:
- with set_forward_context(per_layer_attn_metadata...
```
--------------------------------------------------
mean acceptance length: 2.08
--------------------------------------------------
acceptance at token 0:0.63
acceptance at token 1:0.28
acceptance at token 2:0.11
acceptance at token 3:0.05
acceptance at token 4:0.02
acceptance at token 5:0.00
acceptance at token 6:0.00
```
- with set_forward_context(None...
```
--------------------------------------------------
mean acceptance length: 1.53
--------------------------------------------------
acceptance at token 0:0.40
acceptance at token 1:0.12
acceptance at token 2:0.01
acceptance at token 3:0.00
acceptance at token 4:0.00
acceptance at token 5:0.00
acceptance at token 6:0.00
```

Repro command:
> VLLM_LOGGING_LEVEL=DEBUG VLLM_USE_V1=1 python examples/offline_inference/eagle.py  --num_spec_tokens 7 --max_num_seqs 1 --num_prompts 1 --compilation_config '{"full_cuda_graph": true, "cudagraph_capture_sizes": [1]}'


This seems a bit unexpected, wondering if @zou3519 may have any insights?

Comment (yinghai) [reply vllm/v1/spec_decode/eagle.py]: it does affect things when cudagraph is being captured, right?

Comment (zixi-qi) [reply vllm/v1/spec_decode/eagle.py]: I think capture is supposed to happen inside the dummy_run function instead of here though?

Comment (zou3519) [reply vllm/v1/spec_decode/eagle.py]: Some initial thoughts, if it helps. I'm still digging into this.

- full cuda graphs only applies to the nn.Module that is decorated with support_torch_compile. In eagle, the eagle head gets decorated with support_torch_compile (somewhere the self.model invokes a model, e.g. LlamaModel). The line of code we are commenting on always runs in Python, even when cudagraphs has been recorded.

File: vllm/v1/worker/gpu_model_runner.py [hunk 2]
```diff
@@ -1982,7 +1982,7 @@ def _dummy_run(
 
             if self.speculative_config and self.speculative_config.use_eagle():
                 assert isinstance(self.drafter, EagleProposer)
-                self.drafter.dummy_run(num_tokens)
+                self.drafter.dummy_run(num_tokens, attn_metadata)
```
Comment (zou3519) [inline vllm/v1/worker/gpu_model_runner.py:2188]: Here's my hypothesis:
- the attn_metadata contains tensors
- cudagraphs is baking in the addresses of those tensors
- during runtime, the captured cudagraphs still read from these tensors.

Does the eagle forward pass use the tensors in the attn_metadata? If so, every time we invoke the eagle head, we may need to copy data into the tensors in the attn_metadata.

File: vllm/v1/spec_decode/eagle.py (hunk 1 above)
Comment (zixi-qi) [reply vllm/v1/spec_decode/eagle.py]: Also dug more and found that this line is the "prefill" section for the draft model, where num_tokens can be greater than max configured batch size for cudagraph capture. In this case, the model seems to be running in eager mode which utilizes the passed-in attn_metadata instead of the captured one, which is probably the expected behavior?

File: vllm/v1/worker/gpu_model_runner.py (hunk 2 above)
Comment (zixi-qi) [reply vllm/v1/worker/gpu_model_runner.py:2188]: You are right this is partially the reason for the numerical gap. As an experiment I copied over the attn_metadata constructed for eager mode into the captured attn_metadata in latest commit:

```
# copy attention metadata for full cudagraph mode
if self.draft_attn_metadata is not None:
    self.draft_attn_metadata.seq_lens[:attn_metadata.seq_lens.shape[0]].copy_(attn_metadata.seq_lens.clone())
    self.draft_attn_metadata.slot_mapping[:attn_metadata.slot_mapping.shape[0]].copy_(attn_metadata.slot_mapping.clone())
    self.draft_attn_metadata.query_start_loc[:attn_metadata.query_start_loc.shape[0]].copy_(attn_metadata.query_start_loc.clone())
    self.draft_attn_metadata.block_table[:attn_metadata.block_table.shape[0]].copy_(attn_metadata.block_table.clone())
```

As a result, I got better numerics but there is still a gap comparing with piecewise mode:
- VLLM_USE_V1=1 python examples/offline_inference/eagle.py --num_spec_tokens 7 --num_prompts 1 --compilation_config '{"full_cuda_graph": true, "cudagraph_capture_sizes": [1]}'
```
--------------------------------------------------
mean acceptance length: 2.46
--------------------------------------------------
acceptance at token 0:0.69
acceptance at token 1:0.38
acceptance at token 2:0.20
acceptance at token 3:0.12
acceptance at token 4:0.06
acceptance at token 5:0.00
acceptance at token 6:0.00
```
- VLLM_USE_V1=1 python examples/offline_inference/eagle.py --num_spec_tokens 7 --num_prompts 1 --compilation_config '{"full_cuda_graph": false, "cudagraph_capture_sizes": [1]}'
```
--------------------------------------------------
mean acceptance length: 2.82
--------------------------------------------------
acceptance at token 0:0.77
acceptance at token 1:0.51
acceptance at token 2:0.28
acceptance at token 3:0.13
acceptance at token 4:0.05
acceptance at token 5:0.03
acceptance at token 6:0.03
```

Comment (zixi-qi) [reply vllm/v1/worker/gpu_model_runner.py:2188]: So it seems there might still be some discrepancy in attention computation between eager mode and cudagraph mode. Will try to investigate more and would also appreciate if you have any suggestions to check from torch.compile perspective

File: vllm/v1/spec_decode/eagle.py (hunk 1 above)
Comment (zou3519) [reply vllm/v1/spec_decode/eagle.py]: cc @BoyuanFeng too

Comment (zou3519) [reply vllm/v1/spec_decode/eagle.py]: Offline Zixi and I tried some things:
- turn off inductor
- turn off vllm_compile_cache
- check for dynamic shape issues

All of the above seemed to not do anything, so this is still probably a CUDAGraphs issue

Comment (zixi-qi) [reply vllm/v1/spec_decode/eagle.py]: Did some further investigation and found that changing "use_inductor" settings in compilation config has an impact on the acceptance rate in piecewise cudagraph mode (I initially didn't think this is relevant since the impact to full cudagraph mode is very small, but turns out it impacts the baseline piecewise cudagraph more):

Acceptance rate comparison:
| cudagraph  | use_inductor=True | use_inductor=False | 
| ------------- | ------------- | ------------- |
| full graph | 2.46  | 2.50 |
| piecewise  | 2.82 | 2.50 |

- VLLM_USE_V1=1 python examples/offline_inference/eagle.py --num_spec_tokens 7 --num_prompts 1 --compilation_config '{"full_cuda_graph": false, "cudagraph_capture_sizes": [1], "use_inductor": false}'
```
--------------------------------------------------
mean acceptance length: 2.50
--------------------------------------------------
acceptance at token 0:0.69
acceptance at token 1:0.39
acceptance at token 2:0.22
acceptance at token 3:0.15
acceptance at token 4:0.06
acceptance at token 5:0.00
acceptance at token 6:0.00
```
- VLLM_USE_V1=1 python examples/offline_inference/eagle.py --num_spec_tokens 7 --num_prompts 1 --compilation_config '{"full_cuda_graph": false, "cudagraph_capture_sizes": [1], "use_inductor": true}'
```
--------------------------------------------------
mean acceptance length: 2.82
--------------------------------------------------
acceptance at token 0:0.77
acceptance at token 1:0.51
acceptance at token 2:0.28
acceptance at token 3:0.13
acceptance at token 4:0.05
acceptance at token 5:0.03
acceptance at token 6:0.03
```
- VLLM_USE_V1=1 python examples/offline_inference/eagle.py --num_spec_tokens 7 --num_prompts 1 --compilation_config '{"full_cuda_graph": true, "cudagraph_capture_sizes": [1], "use_inductor": false}'
```
--------------------------------------------------
mean acceptance length: 2.50
--------------------------------------------------
acceptance at token 0:0.69
acceptance at token 1:0.39
acceptance at token 2:0.22
acceptance at token 3:0.15
acceptance at token 4:0.06
acceptance at token 5:0.00
acceptance at token 6:0.00
```
- VLLM_USE_V1=1 python examples/offline_inference/eagle.py --num_spec_tokens 7 --num_prompts 1 --compilation_config '{"full_cuda_graph": true, "cudagraph_capture_sizes": [1], "use_inductor": true}'
```
--------------------------------------------------
mean acceptance length: 2.46
--------------------------------------------------
acceptance at token 0:0.69
acceptance at token 1:0.38
acceptance at token 2:0.20
acceptance at token 3:0.12
acceptance at token 4:0.06
acceptance at token 5:0.00
acceptance at token 6:0.00
```

cc @zou3519 @houseroad

File: vllm/v1/worker/gpu_model_runner.py (hunk 2 above)
Comment (fhl2000) [reply vllm/v1/worker/gpu_model_runner.py:2188]: I am wondering if we can directly reuse the persistent buffer from attnmetadata.

One more thing, I think you should also consider the padding issue if possible.  The buffer from attnmetadata should have correctly filled values for the padding region

Comment (zixi-qi) [reply vllm/v1/worker/gpu_model_runner.py:2188]: Thanks for the suggestion, currently the issue seems to be inductor related: https://github.com/vllm-project/vllm/pull/20190#discussion_r2211819901

File: vllm/v1/spec_decode/eagle.py (hunk 1 above)
Comment (zixi-qi) [reply vllm/v1/spec_decode/eagle.py]: @zou3519 and I further found that if we turn on inductor and only treat rms norm as a custom op, we would also get on par numerics. So it is likely a similar issue as https://github.com/vllm-project/vllm/issues/19403#issuecomment-2958778812
